mmctl tool documentation (#3229)
mattermost/docs: source/administration/mmctl-cli-tool.rst

The mmctl tool is a remote CLI tool for Mattermost which is installed locally and uses the Mattermost API. Authentication
is done with either login credentials or an authentication token.

Being installed locally allows a System Admin to run CLI commands even in instances where there is no access to the
server (e.g., via SSH). This tool is currently in beta and can be used alongside the Mattermost CLI tool.
In the future, the Mattermost CLI tool will be deprecated.

**Notes**

-  Parameters in CLI commands are order-specific.
-  If special characters (``!``, ``|``, ``(``, ``)``, ``\``, ``'``, and ``"``) are used, the entire argument needs to be surrounded by single quotes (e.g. ``-password 'mypassword!'``, or the individual characters need to be escaped out (e.g. `` password mypassword\!``).
- Team name and channel name refer to the handles, not the display names. So in the URL ``https://community.mattermost.com/core/channels/town-square`` team name would be ``core`` and channel name would be ``town-square``.

**Options**

.. code-block:: sh

       --format string    the format of the command output [plain, json] (default "plain")
       -h, --help         help for mmctl

> **Commands**

Reviewer: Missing 'mmctl auth'.

Reviewer: 'version' and 'webhook' can be removed if we decide to remove from this doc for now.

    | Active |        Name | Username |                     InstanceUrl |
    |--------|-------------|----------|---------------------------------|
    |      * | my-instance | john.doe | https://my-instance.example.com |

And now we can run commands normally:

.. code-block:: sh

   $ mmctl user search john.doe
   id: qykfw3t933y38k57ubct77iu9c
   username: john.doe
   nickname:
   position:
   first_name: John
   last_name: Doe
   email: john.doe@example.com
   auth_service:

Installing Shell Completions
^^^^^^^^^^^^^^^^^^^^^^^^^^

To install the shell completions for bash, add the following line to your ``~/.bashrc`` or ``~/.profile`` file:

.. code-block:: sh

  source <(mmctl completion bash)

For zsh, add the following line to your ``~/.zshrc`` file:

.. code-block:: sh

  source <(mmctl completion zsh)

mmctl channel
--------------

Commands for channel management.

  Child Commands
    -  `mmctl channel add`_ - Add a channel
    -  `mmctl channel archive`_ - Archive a channel
    -  `mmctl channel create`_ - Create a channel
    -  `mmctl channel list`_ - List all channels on specified teams
    -  `mmctl channel make_private`_ - Set a channel's type to "private"

>     -  `mmctl channel make`_ - Make a channel

Reviewer: 'make', 'modify', 'move' commands are not in the documentation yet.

.. code-block:: sh

    mmctl post create [flags]

**Examples**

.. code-block:: sh

    post create myteam:mychannel --message "some text for the post"

**Options**

.. code-block:: sh

  -h, --help              help for create
  -m, --message string    Message for the post
  -r, --reply-to string   Post id to reply to

**Options Inherited from Parent Commands**

.. code-block:: sh

    --format string   the format of the command output [plain, json] (default "plain")

mmctl post list
^^^^^^^^^^^^^^^^

**Description**

  List posts for a channel.

**Format**

.. code-block:: sh

   mmctl post list [flags]

**Examples**

.. code-block:: sh

    post list myteam:mychannel
    post list myteam:mychannel --number 20

**Options**

.. code-block:: sh

  -f, --follow       Output appended data as new messages are posted to the channel
  -h, --help         help for list
  -n, --number int   Number of messages to list (default 20)
  -i, --show-ids     Show posts ids

**Options Inherited from Parent Commands**

.. code-block:: sh

    --format string   the format of the command output [plain, json] (default "plain")

mmctl roles
-----------

mmctl team
----------

Management of teams.

> Child Commands

Reviewer: 'archive' and 'rename' can be removed.

.. code-block:: sh

    team list

**Options**

.. code-block:: sh

    -h, --help  help for list

**Options Inherited from Parent Commands**

.. code-block:: sh

    --format string   the format of the command output [plain, json] (default "plain")

mmctl team remove
^^^^^^^^^^^^^^^^^^

**Description**

  Remove specified users from a team.

**Format**

.. code-block:: sh

    mmctl team remove [team] [users] [flags]

**Examples**

.. code-block:: sh

   team remove myteam user@example.com username

**Options**

.. code-block:: sh

    -h, --help  help for remove

**Options Inherited from Parent Commands**

.. code-block:: sh

   --format string   the format of the command output [plain, json] (default "plain")

mmctl team search
^^^^^^^^^^^^^^^^^^

**Description**

  Search for teams based on name.

**Format**

.. code-block:: sh

   mmmctl team search [teams] [flags]

**Examples**

.. code-block:: sh

   team search team1

**Options**

.. code-block:: sh

   -h, --help  help for search

**Options Inherited from Parent Commands**

.. code-block:: sh

   --format string   the format of the command output [plain, json] (default "plain")

mmctl user
---------

Management of users.

> Child Commands

Reviewer: 'user activate' can be removed.

**Options**

.. code-block:: sh

  -a, --access-token string   Access token to use instead of username/password
  -h, --help                  help for login
  -m, --mfa-token string      MFA token for the credentials
  -n, --name string           Name for the credentials
  --no-activate               If present, it won't activate the credentials after login
  -p, --password string       Password for the credentials
  -u, --username string       Username for the credentials

**Options Inherited from Parent Commands**

.. code-block:: sh

   --format string the format of the command output [plain, json] (default "plain")

mmctl auth renew
^^^^^^^^^^^^^^^^^

**Description**

  Renew the credentials for a given server.

**Format**

.. code-block:: sh

   mmctl auth renew [flags]

**Examples**

.. code-block:: sh

   auth renew local-server

**Options**

.. code-block:: sh

  -a, --access-token string   Access token to use instead of username/password
  -h, --help                  help for renew
  -m, --mfa-token string      MFA token for the credentials
  -p, --password string       Password for the credentials

> Options Inherited from Parent Commands

Reviewer: ```suggestion
**Options Inherited from Parent Commands**
```

.. code-block:: sh

   --format string the format of the command output [plain, json] (default "plain")

mmctl auth set
^^^^^^^^^^^^^^^^^

**Description**

  Set credentials to use in the following commands.

**Format**

.. code-block:: sh

   mmctl auth set [server name] [flags]

**Examples**

.. code-block:: sh

   auth set local-server

**Options**

.. code-block:: sh

   -h, --help   help for set

**Options Inherited from Parent Commands**

.. code-block:: sh

   --format string the format of the command output [plain, json] (default "plain")

> Authenticate to a server (e.g. >mmctl auth login https://test.mattermost.com). Then username and password

Reviewer: ```suggestion
Authenticate to a server (e.g. >mmctl auth login https://test.mattermost.com), then username and password
```

Not sure if keeping the sentences separate is intentional.

(and MFA token if MFA is enabled on the account).

Password

.. code-block:: sh

     $ mmctl auth login https://community.mattermost.com --name community --username my-username --password mysupersecret

The ``login`` command can also work interactively, so if you leave any required flag empty, ``mmctl`` will ask you for it interactively:

.. code-block:: sh

    $ mmctl auth login https://community.mattermost.com
    Connection name: community
    Username: my-username
    Password:

MFA

To log in with MFA, use the ``--mfa-token`` flag:

.. code-block:: sh

   $ mmctl auth login https://community.mattermost.com --name community --username my-username --password mysupersecret --mfa-token 123456

Access Tokens
^^^^^^^^^^^^^

You can generate and use a personal access token to authenticate with a server, instead of using username and password to log in:

.. code-block:: sh

   $ mmctl auth login https://community.mattermost.com --name community --access-token MY_ACCESS_TOKEN

> into a Mattermost instance:

Reviewer: Could be intentional but this sentence seems incomplete.

Author: Fixed.

.. code-block:: sh

   --format string   the format of the command output [plain, json] (default "plain")

mmctl channel list
^^^^^^^^^^^^^^^^^

**Description**

  List all channels on specified teams. Archived channels are appended with '(archived)'.

**Format**

.. code-block:: sh

   mmctl channel list [teams] [flags]

**Examples**

.. code-block:: sh

  channel list myteam

**Options**

.. code-block:: sh

  -h, --help   help for list

**Options Inherited from Parent Commands**

.. code-block:: sh

   --format string   the format of the command output [plain, json] (default "plain")

mmctl channel make_private
^^^^^^^^^^^^^^^^^^^^^^^^^^

**Description**

   Set the type of a channel from public to private. Channel can be specified by ``[team]:[channel]`` (i.e., myteam:mychannel)
   or by channel ID.

**Format**

.. code-block:: sh

    mmctl channel make_private [channel] [flags]

**Examples**

.. code-block:: sh

    channel make_private myteam:mychannel

**Options**

.. code-block:: sh

  -h, --help   help for make_private

**Options Inherited from Parent Commands**

.. code-block:: sh

 --format string   the format of the command output [plain, json] (default "plain")

> mmctl channel make

Reviewer: These are marked as in progress. May be intentional.

  -i, --show-ids     Show posts ids

**Options Inherited from Parent Commands**

.. code-block:: sh

    --format string   the format of the command output [plain, json] (default "plain")

mmctl roles
-----------

mmctl team
----------

Management of teams.

Child Commands
  -  `mmctl team add`_ - Add teams
  -  `mmctl team archive`_ - Archive teams
  -  `mmctl team create`_ - Create teams
  -  `mmctl team delete`_ - Delete teams
  -  `mmctl team list`_ - List teams
  -  `mmctl team remove`_ - Remove teams
  -  `mmctl team rename`_ - Rename teams
  -  `mmctl team search`_ - Search teams

**Options**

.. code-block:: sh

  -h, --help   help for team

mmctl team add
^^^^^^^^^^^^^^^^

**Description**

  Add specified users to a team.

**Format**

.. code-block:: sh

    mmctl team add [team] [users] [flags]

**Examples**

.. code-block:: sh

    team add myteam user@example.com username

**Options**

.. code-block:: sh

   -h, --help   help for add

**Options Inherited from Parent Commands**

.. code-block:: sh

    --format string   the format of the command output [plain, json] (default "plain")

mmctl team archive
^^^^^^^^^^^^^^^^

> *In progress*

Reviewer: In progress

Author: Same as previous comment - I worked off the migration sheet and the repo and thought it might be helpful to list them all but mark which ones are still being worked on. Can definitely remove this.

Author: Removed.

   team remove myteam user@example.com username

**Options**

.. code-block:: sh

    -h, --help  help for remove

**Options Inherited from Parent Commands**

.. code-block:: sh

   --format string   the format of the command output [plain, json] (default "plain")

mmctl team search
^^^^^^^^^^^^^^^^^^

**Description**

  Search for teams based on name.

**Format**

.. code-block:: sh

   mmmctl team search [teams] [flags]

**Examples**

.. code-block:: sh

   team search team1

**Options**

.. code-block:: sh

   -h, --help  help for search

**Options Inherited from Parent Commands**

.. code-block:: sh

   --format string   the format of the command output [plain, json] (default "plain")

mmctl user
---------

Management of users.

Child Commands
  -  `mmctl user activate`_ - Activate user
  -  `mmctl user create`_ - Create user
  -  `mmctl user deactivate`_ - Deactivate user
  -  `mmctl user email`_ - Set user email
  -  `mmctl user invite`_ - Invite user
  -  `mmctl user reset_password`_ - Reset user password
  -  `mmctl user resetmfa`_ - Reset user's MFA token
  -  `mmctl user search`_ - Search for a user

**Options**

.. code-block:: sh

   -h, --help       help for user

> mmctl user activate

Reviewer: Code blocks in "mmctl user activate" section are not rendering correctly.

.. code-block:: sh

  --format string   the format of the command output [plain, json] (default "plain")

mmctl user resetmfa
^^^^^^^^^^^^^^^^^^^^

**Description**

  Turn off multi-factor authentication for a user. If MFA enforcement is enabled, the
  user will be forced to re-enable MFA as soon as they login.

**Format**

.. code-block:: sh

    mmctl user resetmfa [users] [flags]

**Examples**

.. code-block:: sh

    user resetmfa user@example.com

**Options**

.. code-block:: sh

    -h, --help       help for resetmfa

**Options Inherited from Parent Commands**

.. code-block:: sh

  --format string   the format of the command output [plain, json] (default "plain")

mmctl user search
^^^^^^^^^^^^^^^^^^

**Description**

  Search for users based on username, email, or user ID.

**Format**

.. code-block:: sh

    mmctl user search [users] [flags]

**Examples**

.. code-block:: sh

    user search user1@mail.com user2@mail.com

**Options**

.. code-block:: sh

    -h, --help       help for search

**Options Inherited from Parent Commands**

.. code-block:: sh

  --format string   the format of the command output [plain, json] (default "plain")

> mmctl version

Reviewer: Few sections here are incomplete. May be intentional

Author: Removed.

mmctl Command Line Tool (Beta)
==============================

The mmctl tool is a remote CLI tool for Mattermost which is installed locally and uses the Mattermost API. Authentication

> is done with either login credentials or an authentication token. <elaborate more on this, provide links to other docs>.

Reviewer: There is an internal note here to elaborate / add links

Author: Thanks for catching that!

Being installed locally allows a System Admin to run CLI commands even in instances where there is no access to the
server (e.g., via SSH). This tool is currently in beta and can be used alongside the Mattermost CLI tool.
In the future, the Mattermost CLI tool will be deprecated.

**Notes**

 -  Parameters in CLI commands are order-specific.

>  -  If special characters (``!``, ``|``, ``(``, ``)``, ``\``, ``'``, and ``"``) are used, the entire argument needs to be surrounded

Reviewer: Some of these bullet points need formatting tweaks.

Author: Thanks! Fixed.

 by single quotes (e.g. ``-password 'mypassword!'``, or the individual characters need to be escaped out (e.g. ``-password mypassword\!``).
 -  Team name and channel name refer to the handles, not the display names. So in the URL ``https://community.mattermost.com/core/channels/town-square`` team
 name would be ``core`` and channel name would be ``town-square``.

Options

> .. code-block:: sh

Reviewer: This table might need formatting fixes.

mmctl Command Line Tool (Beta)
==============================

The mmctl tool is a remote CLI tool for Mattermost which is installed locally and uses the Mattermost API. Authentication
is done with either login credentials or an authentication token.

Being installed locally allows a System Admin to run CLI commands even in instances where there is no access to the
server (e.g., via SSH). This tool is currently in beta and can be used alongside the Mattermost CLI tool.
In the future, the Mattermost CLI tool will be deprecated.

This feature was developed to a large extent by community contributions and we'd like to extend our gratitude to the contributors who have

> worked on this project. We are currently accepting PRs for Help Wanted issues

Reviewer: ```suggestion
worked on this project. We are currently accepting pull requests for Help Wanted issues
```

Reviewer: This one is not fixed anymore.

Author: That's my bad - I thought I'd committed it and must have over-written the change. Thanks!

- Team name and channel name refer to the handles, not the display names. So in the URL ``https://community.mattermost.com/core/channels/town-square`` team name would be ``core`` and channel name would be ``town-square``.

**Options**

.. code-block:: sh

       --format string    the format of the command output [plain, json] (default "plain")
       -h, --help         help for mmctl

**Commands**
   - `mmctl auth`_ - Authentication Management
   - `mmctl channel`_ - Channel Management
   - `mmctl command`_ - Command Management
   - `mmctl config`_ - Configuration Management
   - `mmctl group`_ - Group Management
   - `mmctl ldap`_ - LDAP Management
   - `mmctl license`_ - License Management
   - `mmctl logs`_ - Log Management
   - `mmctl permissions`_ - Permissions Management
   - `mmctl plugin`_ - Plugin Management
   - `mmctl post`_ - Post Management
   - `mmctl roles`_ - Roles Management
   - `mmctl team`_ - Team Management
   - `mmctl user`_ - User Management
   - `mmctl version`_ - Version Management
   - `mmctl webhook`_ - Webhook Management

Installing mmctl
----------------

> To install the project in your `$GOPATH` run the following command as admin user:

Reviewer 2: We can include here the use of `brew install mmctl` command too. What do you think @mgdelacroix ?

Author: This is a good suggestion. Would you suggest breaking this into sub-sections for different OSs or including it as an alternative in a single sentence?

Reviewer 2: That would be great. We could do it by OS or we can have two sections one for building the tool and the other one to install from package manager

Reviewer 3: I agree that separating the build and the install process would be good. I think that for the different install processes we can create a "Methods" section, and then describe ways to install it "via go get", "via brew", "via release package", etc. Chances are that we end up having more package managers to add here, and as some OSs share install methods (release and go get are for all of them, brew is for OSX and Linux...) I think having them under "Methods" suits it better

Author: Great, thank you! Would you be open to sharing the methods and I can add them to the doc and do the restructuring?

Reviewer 3: Sure, right now we have:

### Via brew

Valid for Linux and Mac, with brew installed, running `brew install mmctl` will install the tool.

### Via go get

Valid for Linux, Mac and Windows, with the go environment configured, running `go get -u github.com/mattermost/mmctl` would install the tool.

### Via release package

Valid for Linux, Mac and Windows, user needs to go to the [mmctl releases page](https://github.com/mattermost/mmctl/releases) and download the appropriate release for their OS, which will contain the binary of the tool.

Author: I've updated the doc with this feedback.

Installing mmctl
----------------

There are different methods available to install mmctl.

**Via brew (Linux, macOS)**

Use this option on Linux and macOS if you have Homebrew installed.

.. code-block:: sh

   brew install mmctl

**Via go get (Linux, macOS, Windows)**

> Use this option on Linux, macOS, and Windows if you have a go environment configured.

Reviewer 3: The `$GOPATH` environment variable reference belongs to this method:

```suggestion
Use this option on Linux, macOS, and Windows if you have a go environment configured. This would install the project in your `$GOPATH`:
```

Author: Changed it slightly differently. Let me know if it doesn't work.

Reviewer 3: @justinegeffen just one nit, you don't need to be the admin user to install the tool, just to have the go environment set up.

Author: Fixed, thank you!

.. code-block:: sh

   go get -u github.com/mattermost/mmctl

**Via release package (Linux, macOS, Windows)**

Vist the `mmctl releases page <https://github.com/mattermost/mmctl/releases>`__ and download the appropriate release for your OS, and install the binary.

To add the project in your `$GOPATH` from the package manager run the following command as admin user:

.. code-block:: sh

>    go get -u github.com/mattermost/mmctl

Reviewer 3: This was covered in the "via go get" method, so we can safely remove it
